### PR TITLE
Widget Catalog

### PR DESCRIPTION
**Pull request by FaisalAbid**
Fixes https://github.com/flutter/flutter/issues/9531

### CODE REVIEW

**Comment by FaisalAbid** [comment]
https://www.dropbox.com/s/yc2s5nxww0fzcif/demo.mov?dl=0

screenshot 2017-05-08 02 32 21
screenshot 2017-05-08 02 32 25
screenshot 2017-05-08 02 33 32
screenshot 2017-05-08 02 33 36

File: widgets/catalog/materialdesign.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  - /css/catalog.css
+---
+<div class="catalog">
+{% for section in site.data.catalog.index %}
+ {% if section.name contains 'Material Design' %}
+ <div class="category-description"><p>{{section.description}}</p></div>
+ {% endif %}
+{% endfor %}
+
+<!-- Data for the catalog comes from widgets.csv in the _data folder -->
```
Comment (InMatrix) [inline widgets/catalog/materialdesign.md]: This comment is old, since we no longer use csv?

File: _data/catalog/index.json [hunk 2]
```diff
@@ -0,0 +1,114 @@
+[
+  {
+    "name": "Basics",
+    "description": "Widgets you absolutely need to know before building your first Flutter app.",
+    "priority": "high",
+    "id":"basics"
+  },
+  {
+    "name": "Material Design",
+    "description": "Flutter implements many of the UI components from Google’s Material Design guidelines.",
```
Comment (InMatrix) [inline _data/catalog/index.json]: Can a category's description contain links? For example, I'd like to link "Material Design guidelines" to material.io/guidelines

Comment (FaisalAbid) [inline _data/catalog/index.json]: Yes I belive that will work, I will do a quick test and check shortly

**Comment by sethladd** [comment]
* How does this look on a phone / narrow browser window?

* Let's try to use a grid for the individual widget cards. Instead of wide, row-by-row listings under Basic, let's see how to make this more info-dense.

Thanks Faisal!

**Comment by FaisalAbid** [comment]
@sethladd 
screenshot 2017-05-08 11 45 55
screenshot 2017-05-08 11 44 13

**Comment by FaisalAbid** [comment]
So basically instead of the rows you want the info to be in a grid?

**Comment by sethladd** [comment]
> So basically instead of the rows you want the info to be in a grid?

I think we should try to reduce the amount of up-and-down scrolling, so maybe at least a two-column grid?

**Comment by FaisalAbid** [comment]
Okay got it! Will try to do it asap in the next hour or so

**Comment by sethladd** [comment]
Noticed: broken image here:

screen shot 2017-05-08 at 3 29 15 pm

**Comment by sethladd** [comment]
No image here:
screen shot 2017-05-08 at 3 29 38 pm

**Comment by sethladd** [comment]
cards don't quite shrink correctly:

screen shot 2017-05-08 at 3 42 26 pm

**Comment by sethladd** [comment]
when in single column, I see different widths:

screen shot 2017-05-08 at 3 42 58 pm

File: _includes/catalogpage.html [hunk 3]
```diff
@@ -0,0 +1,61 @@
+<div class="catalog">
+    {% for section in site.data.catalog.index %}
+    {% if section.name contains include.category %}
+    <div class="category-description"><p>{{section.description}}</p></div>
+    {% endif %}
+    {% endfor %}
+
+    <ul>
+        {% for category in site.data.catalog.index %}
+        {% if category.name == 'Async' %}
```
Comment (sethladd) [inline _includes/catalogpage.html]: are you sure you want this here?

File: _data/catalog/widgets.json [hunk 4]
```diff
@@ -0,0 +1,1569 @@
...
+    "description": "A convenience widget that wraps a number of widgets that are commonly required for material design applications.",
+    "categories": [],
+    "subcategories": [
+      "App structure and navigation"
+    ],
+    "link": "https://docs.flutter.io/flutter/material/MaterialApp-class.html",
+    "image": "https://storage.googleapis.com/material-design/publish/material_v_11/assets/0Bx4BSt6jniD7Y1huOXVQdlFPMmM/materialdesign_introduction.png"
+  },
+  {
+    "name": "WidgetsApp",
```
Comment (InMatrix) [inline _data/catalog/widgets.json]: WidgetsApp is not really a Material Design widget. Let's move it to "See also" at the end of this sub-category, once we have the ability to do so.

**Comment by sethladd** [comment]
@mit-mit do you know how we can tell the link checker not to traverse out to https://flutter.io when testing a PR? It will fail when the deployed website doesn't have a new page, which only exists in the PR. Thanks!

**Comment by mit-mit** [comment]
The link checker already replaces the https://flutter.io prefix away from the canonical urls:
https://github.com/flutter/website/blob/master/Rakefile#L16

Unfortunately the error output it prints still contains `https://flutter.io`, but you just ignore that.

I currently see just two failures:
```
- ./_site/widgets-intro/index.html
  *  linking to https://flutter.io/widgets/#layout-models, but layout-models does not exist (line 214)
     <a href="https://flutter.io/widgets/#layout-models">Widgets Overview - Layout Models</a>
  *  linking to https://flutter.io/widgets/#material-design-widgets, but material-design-widgets does not exist (line 379)
     <a href="https://flutter.io/widgets/#material-design-widgets">Widgets Overview - Material Design Widgets</a>
rake aborted!
HTML-Proofer found 2 failures!
```

So what these are actually saying are:

```
- ./_site/widgets-intro/index.html
  *  linking to /widgets/#layout-models, but layout-models does not exist (line 214)
  *  linking to /widgets/#material-design-widgets, but material-design-widgets does not exist (line 379)
```

If you look in widgets.md (or I guess rather /_data/catalog/index.json) those `id`'s are indeed not present!

**Comment by FaisalAbid** [comment]
I fixed the broken links @sethladd 

now we're getting an undefined error at Pathprovider. taking a look

**Comment by mit-mit** [comment]
@FaisalAbid PathProvider issue is related to a flutter/flutter change we just landed! I'll get that fixed right away, stay tuned!

**Comment by FaisalAbid** [comment]
ah thank god, I was so confused. Thanks @mit-mit

**Comment by sethladd** [comment]
@mit-mit thanks! that definitely helps explain it. I, too, was confused :)

**Comment by mit-mit** [comment]
@FaisalAbid I merged https://github.com/flutter/website/pull/548, so try to merge with master, and hopefully this should pass validation

**Comment by FaisalAbid** [comment]
@mit-mit ok thanks!

**Comment by sethladd** [comment]
lgtm

thanks!

**Comment by Hixie** [comment]
Should we remove the old .csv files now?

**Comment by sethladd** [comment]
Good catch, thanks!

**Comment by Hixie** [comment]
k, i'll take care of it

**Pull request merged**